zebra/if_netlink.c

```diff
@@ -412,14 +413,20 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 				ifname, errno, safe_strerror(errno));
 		/* no device means interface unreachable */
 		if (errno == ENODEV && error)
-			*error = -1;
+			*error = INTERFACE_SPEED_ERROR_READ;
 		ecmd.speed_hi = 0;
 		ecmd.speed = 0;
 	}
 
 	close(sd);
 
-	return ((uint32_t)ecmd.speed_hi << 16) | ecmd.speed;
+	ret = ((uint32_t)(ecmd.speed_hi << 16)) | ecmd.speed;
```

Comment:
This is still wrong. You are still shifting into the sign bit on a signed integer before you cast. You need to keep the cast exactly as it was.

Reply:
redone.